File: epi_judge_cpp/number_of_score_combinations.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

int CalculateFinalCombination(int final_score, const vector<int>& individual_play_scores, 
                              vector<vector<int>>& num_combination){
    for (int i = 0; i < individual_play_scores.size(); i++) {
        num_combination[i][0] = 1;
        for (int j = 1; j <= final_score; j++) {
            int without_own = i >= 1 ? num_combination[i - 1][j]:0;
            int with_own = j >= individual_play_scores[i] ? num_combination[i][j - individual_play_scores[i]] : 0;
            num_combination[i][j] = without_own + with_own;
        }
    }
    return num_combination.back().back();
}

int NumCombinationsForFinalScore(int final_score,
                                 const vector<int>& individual_play_scores) {
    vector<vector<int>> num_combination(individual_play_scores.size(), vector<int>(final_score+1, 0));
    return CalculateFinalCombination(final_score, individual_play_scores,num_combination);
}


int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"final_score", "individual_play_scores"};
  return GenericTestMain(args, "number_of_score_combinations.cc",
                         "number_of_score_combinations.tsv",
                         &NumCombinationsForFinalScore, DefaultComparator{},
                         param_names);
}

File: epi_judge_python/number_of_score_combinations.py
from test_framework import generic_test


def num_combinations_for_final_score(final_score, individual_play_scores):
    A = [[0] * (final_score + 1) for _ in individual_play_scores]
    for r in A:
        r[0] = 1
    for i in range(len(A)):
        for j in range(1, final_score + 1):
            without_play = A[i-1][j] if i > 0 else 0
            with_play = A[i][j-individual_play_scores[i]] if j >= individual_play_scores[i] else 0
            A[i][j] = without_play + with_play
    return A[-1][-1]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("number_of_score_combinations.py",
                                       "number_of_score_combinations.tsv",
                                       num_combinations_for_final_score))
